docs/source/getting-started/assessing-utility/composition.ipynb

```diff
@@ -0,0 +1,742 @@
+{
```

Comment:
One thing I will say about this. It's not fully clear through this doc what the differences are among interactive, sequential, and adaptive. Can these three things be true about a set of queries simultaneously? Can a subset of these three be true for a set of queries? Which combinations are valid. It looks a bit mixed up right now, unfortunately.

Reply:
yeah, currently i go into all of them a bit in this introduction suggestion. what combinations are valid and invalid? I could add that in

Reply:
My sense is that 
- *mechanisms* can be described as interactive or non-interactive, 
-  but *compositions* can be described as adaptive or non-adaptive.

(Basing this on pages 104 and 151 in _Hands-on DP_.)

So, in the discussion here, maybe we only mention adaptive/non-adaptive compositions, and not mention the parallel distinction between  interactive/non-interactive mechanisms?

Reply:
From Mike:

> Mechanisms can be interactive or non-interactive. The difference between them is that interactivity allows an adversary to gain more information via communication with the output. We technically always consider measurements in the library to be interactive mechanisms, and in the case where the mechanism would typically be considered non-interactive, the adversary is trivial (there are zero rounds of communication). Informally, a mechanism is non-interactive if there is no queryable in the output (nothing to communicate with).
>
> Compositors are a kind of mechanism. Compositors can also be either interactive or non-interactive. Some examples:
> - non-interactive: `make_basic_composition` (for each privacy measure), advanced composition (for pure/approx/BR-DP)
> - interactive: `make_sequential_composition` (likewise), advanced composition (likewise)
> You can even have non-interactive composition of interactive mechanisms (for example `make_basic_composition` over a bunch of interactive mechanisms).
>
> If the composition is interactive, then it may either be adaptive or concurrent.
> - If the compositor is adaptive, then there is a sequentiality constraint on queries to interactive mechanisms spawned by the compositor: only the most recent queryable spawned may be interacted with.
> - If the compositor is concurrent, then the adversary may submit multiple interactive mechanisms to the compositor, thus spawning multiple queryables, and interleave queries amongst them.
> 
> ### Book vs. Library
> - In the book, I put these three cases along one axis of "query interactivity": a compositor is either non-adaptive (corresponding to non-interactive), adaptive, or concurrent
> - In the library, I've hidden the distinction between adaptive and concurrent composition. Concurrent composition allows for everything adaptive composition does and more. So if the user wants an interactive compositor, they simply use `make_sequential_composition`, and the library allows concurrency if it has been proven. If the privacy measure doesn't support concurrent composition, then the implementation imposes a sequentiality constraint to get "adaptive" composition. If we later prove that a definition of privacy allows for concurrent composition, we can then lift the sequentiality constraint for that privacy measure without requiring any changes from the user.